text/0143-sdk-fail-safe-mode.md

```diff
@@ -0,0 +1,302 @@
+| 3.2. | _[continued from 3.]_ and __can__ create and send some SDK crash reports. | Crashes and some data. |
+| 4. | The SDK continuously crashes when creating most crash reports, so there is no crash report. | No data. |
+| | __User's Application__ | |
+| 5. | The user's application crashes shortly after the initialization of the Sentry SDK. | Potentially incorrectly disabling the SDK and no data. |
+| 6. | The user's application crashes async during the initialization of the Sentry SDK _[continue at 6.1, 6.2, 6.3]_ |  |
+| 6.1. | _[continued from 6.]_ and the Sentry SDK can write a crash report, which it creates and sends on the next launch. | Potentially incorrectly disabling the SDK and no data. |
+| 6.2. | _[continued from 6.]_ and the Sentry SDK can write a crash report, which it creates but can't send on the next launch. | There isn't much we can do about this, except educating our customers about the importance of initializing the Sentry SDK as early as possible. Even if we incorrectly disable the SDK, it makes no difference. |
+| 6.3. | _[continued from 6.]_ and the Sentry SDK can't write a crash report, because it happens before initializing the crash handlers. | Same as 6.2. |
+| 7. | The user's application crashes before the initialization of the Sentry SDK. | Same as 6.2. |
+| | __Hybrid SDKs__ | |
+| 8. | The hybrid SDK crashes during its initialization, _[continue at 8.1, 8.2]_ |  |
+| 8.1. | _[continued from 8.]_ and the underlying native SDK __can__ initialize. |  |
+| 8.2. | _[continued from 7.]_ and the underlying native SDK __can't__ initialize. |  |
+
+## Option 1: Checkpoints <a name="option-1"></a>
+
+The SDK stores checkpoints via marker files to disk to identify if it completes SDK initialization. When it doesn't, the SDK disables itself. The SDKs should use marker files because checking the file's existence is significantly more performant than reading its contents. To be aware of when the SDK disables itself, we could implement this option in combination with the [Option 3: Failing SDK Endpoint](#option-3).
+
+The SDK implements a retry logic to minimize the risk of wrongly disabling itself. When the app launched x times, the SDK retries if it can launch successfully. If it does, it goes back to normal. If it doesn't, it exponentially increases the number of app launches until it retries.
+
+The specification is written in the [Gherkin syntax](https://cucumber.io/docs/gherkin/reference/). The specification might not work for all edge cases yet, as it can be complicated to get it right. We'll figure out the exact details once we decide to implement it, but it should cover the main scenarios.
+
+```Gherkin
+Scenario: SDK version inits the first time
+    Given there is no launch marker file for the SDK version
+    And there is no success init marker file for the SDK version
+    When the SDK inits
+    Then the SDK creates a launch marker file for the SDK version
```

Comment:
m: The only thing I want to suggest is to not use more than one file. The SDK is either in a broken state (file exists), or not (file does not exist). After a successful initialization, we just delete the file marker.

Reply:
fwiw the previous implementation of this i built just used user defaults. one key, whatever value you need to represent the last state reached during initialization.

Reply:
Im my tests, `UserDefaults` is no longer (Apple changed something at some point) fast enough. It has a save process that is no synchronous anymore.

Reply:
Thanks for pointing that out; I already included this in the RFC: https://github.com/getsentry/rfcs/blob/a66a5b1747712ede252c4222b58ddffe564ee047/text/0143-sdk-fail-safe-mode.md#L167